app/forms/submissions/base_validator.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+module Submissions
+  class BaseValidator
+    include ActiveModel::Model
+
+    class_attribute :form_validators, instance_writer: false, default: {}
+
+    class << self
+      def trn_validator(name, options)
+        form_validators[name] = options
+      end
+    end
+
+    trn_validator :personal_details, form: "PersonalDetailsForm", unless: :apply_application?
+    trn_validator :contact_details, form: "ContactDetailsForm", unless: :apply_application?
+    trn_validator :diversity, form: "Diversities::FormValidator", unless: :apply_application?
+    trn_validator :degrees, form: "DegreesForm", if: :validate_degree?
+    trn_validator :course_details, form: "CourseDetailsForm"
+    trn_validator :training_details, form: "TrainingDetailsForm"
+    trn_validator :trainee_data, form: "ApplyApplications::TraineeDataForm", if: :apply_application?
+    trn_validator :schools, form: "Schools::FormValidator", if: :requires_schools?
+    trn_validator :funding, form: "Funding::FormValidator"
+
+    delegate :requires_schools?, :requires_degree?, :apply_application?, to: :trainee
+
+    validate :submission_ready
+
+    def initialize(trainee:)
+      @trainee = trainee
+    end
+
+    def validate_degree?
+      !apply_application? && requires_degree?
+    end
+
+  private
+
+    attr_reader :trainee
+
+    def progress_keys
```

Comment:
I wonder whether we should call this `validator_keys` rather than anything to do with progress, since it's used in the `MissingDataValidator`.